sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningDeserializer.java

```diff
@@ -65,58 +68,185 @@ public static SimpleModule getModule(final ObjectMapper mapper) {
         SimpleModule module = new SimpleModule();
         module.setDeserializerModifier(new BeanDeserializerModifier() {
             @Override
-            public JsonDeserializer<?> modifyDeserializer(DeserializationConfig config, BeanDescription beanDesc,
+            public JsonDeserializer<?> modifyDeserializer(DeserializationConfig config,
+                                                          BeanDescription beanDesc,
                                                           JsonDeserializer<?> deserializer) {
-                if (beanDesc.getBeanClass().getAnnotation(JsonFlatten.class) != null) {
+                if (BeanDeserializer.class.isAssignableFrom(deserializer.getClass())) {
+                    // Register 'FlatteningDeserializer' for complex type so that 'deserializeWithType'
+                    // will get called for complex types and it can analyze typeId discriminator.
                     return new FlatteningDeserializer(beanDesc.getBeanClass(), deserializer, mapper);
+                } else {
+                    return deserializer;
                 }
-                return deserializer;
             }
         });
         return module;
     }
 
     @SuppressWarnings("unchecked")
     @Override
-    public Object deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
-        JsonNode root = mapper.readTree(jp);
+    public Object deserializeWithType(JsonParser jp,
+                                      DeserializationContext cxt,
+                                      TypeDeserializer tDeserializer) throws IOException {
+        // This method will be called from Jackson for each "Json object with TypeId" as it
+        // process the input data. This enable us to pre-process then give it to the next
+        // deserializer in the Jackson pipeline.
+        //
+        // The parameter 'jp' is the reader to read "Json object with TypeId"
+        //
+        JsonNode currentJsonNode = mapper.readTree(jp);
```

Comment:
Should we just cast this to `ObjectNode` ahead of time so we don't need to cast all the time later in the method?

Reply:
Please see the below [comment](https://github.com/Azure/azure-sdk-for-java/pull/10310#discussion_r410349470) on upcasting based on the presence of `jsonProperty`, similar constraint applied here but based on the presence of `JsonTypeInfo`.